lib/IRGen/GenType.cpp

```diff
@@ -1797,6 +1797,57 @@ static bool isIRTypeDependent(IRGenModule &IGM, NominalTypeDecl *decl) {
   }
 }
 
+bool irgen::mightContainMetadata(const IRGenModule &IGM, const CanType type) {
+  IRGenModule &IGMCast = const_cast<IRGenModule &>(IGM);
+  auto runType = irgen::getRuntimeReifiedType(IGMCast, type);
+  if (isa<SILFunctionType>(runType)) {
```

Comment:
What's special about SILFunctionTypes here?

Reply:
You can't emit metadata type ref for SILFunctionTypes. Lets say your archetype is a tuple containing a SILFunctionType: the way metadata is emitted for tuples is that you emit for all fields regardless of what they are. the code will crash when trying to emit for SILFunctionTypes

Reply:
This suggests the whole approach is wrong, because you're looking at lowered types and not formal types. You cannot emit metadata for lowered types in general, even if it looks like a formal type.

You should not see SILFunctionTypes (or SILBoxTypes), or Metatypes with a representation set, etc, here at all.

Reply:
Basically, if you're calling mightContainMetadata() with some SILType's T.getSwiftRValueType(), you should figure out why and try not to do that.

Reply:
The thing Is the caller function got a cache from the lowered types to previously emitted meta type information, so emit metadata operation will be a nop (which Is what I want to happen for code size and performance reasons) - calling emit metadata on the formal type will not re-use the cache

Reply:
The cache should use formal types as the key, then.